File: temas/herencia/Python/Libro.py
class Libro:
    def __init__(self, titulo, autor, numPaginas):
        self.titulo = titulo
        self.autor = autor
        self.numPaginas = numPaginas

    def imprimir(self):
        print(f"El título es: {self.titulo}\nEscrito por: {self.autor}\nCuenta con un total de {self.numPaginas} páginas\n")

class Novela(Libro):
    def __init__(self, titulo, autor, numPaginas, genero):
        super().__init__(titulo, autor, numPaginas)
        self.genero = genero

    def imprimir(self):
        print(f"El título es: {self.titulo}\nEscrito por: {self.autor}\nCuenta con un total de {self.numPaginas} páginas\nSu género es {self.genero}\n")

# Ejemplo de uso
miNovela = Novela("Antes de Diciembre", "Joana Marcús", 493, "Romántico juvenil")
miNovela2 = Novela("El mapa de los anhelos", "Alice Kellen", 496, "Romántica contemporánea")
miNovela.imprimir()
miNovela2.imprimir()

File: temas/herencia/Scala/Libro.scala
abstract class Libro(val titulo: String, val autor: String) {
  def leer(): Unit
}

class LibroRomance(titulo: String, autor: String, val genero: String) extends Libro(titulo, autor) {
  override def leer(): Unit = {
    println(s"Leyendo $titulo de $autor del género $genero")
  }
}

class LibroClasificado(titulo: String, autor: String, val edad: String) extends Libro(titulo, autor) {
  override def leer(): Unit = {
    println(s"Leyendo $titulo escrito por $autor con una clasificacion de edad $edad")
  }
}

//Ejemplo de uso

object Main {
  def main(args: Array[String]): Unit = {
    val libro1 = new LibroRomance("El mapa de los anhelos", "Alice Kellen", "Romance contemporáneo")
    val libro2 = new LibroClasificado("El principito", "Antoine de Saint-Exupéry", " 6-7 años")
    
    libro1.leer() 
    libro2.leer() 
  }
}
